Implement basic clustered decal projectors.

pcwalton opened the pull request:
This commit adds support for *decal projectors* to Bevy, allowing for textures to be projected on top of geometry. Decal projectors are clusterable objects, just as punctual lights and light probes are. This means that decals are only evaluated for objects within the conservative bounds of the projector, and they don't require a second pass.

These clustered decals require support for bindless textures and as such currently don't work on WebGL 2, WebGPU, macOS, or iOS. For an alternative that doesn't require bindless, see PR #16600. I believe that both contact projective decals in #16600 and clustered decals are desirable to have in Bevy. Contact projective decals offer broader hardware and driver support, while clustered decals don't require the creation of bounding geometry.

A new example, `decal_projectors`, has been added, which demonstrates multiple decals on a rotating object. The decal projectors can be scaled and rotated with the mouse.

There are several limitations of this initial patch that can be addressed in follow-ups:

1. There's no way to specify the Z-index of decals. That is, the order in which multiple decals are blended on top of one another is arbitrary. A follow-up could introduce some sort of Z-index field so that artists can specify that some decals should be blended on top of others.

2. Decals don't take the normal of the surface they're projected onto into account. Most decal implementations in other engines have a feature whereby the angle between the decal projector and the normal of the surface must be within some threshold for the decal to appear. Often, artists can specify a fade-off range for a smooth transition between oblique surfaces and aligned surfaces.

3. There's no distance-based fadeoff toward the end of the projector range. Many decal implementations have this.

This addresses #2401.
 
## Showcase

![Screenshot 2025-01-11 052913](https://github.com/user-attachments/assets/8fabbafc-60fb-461d-b715-d7977e10fe1f)

JMS55 commented on crates/bevy_pbr/src/decal/projector.rs:
```diff
@@ -0,0 +1,364 @@
+    /// The image that the decal projector projects.
+    ///
+    /// This must be a 2D image. If it has an alpha channel, it'll be alpha
+    /// blended with the underlying surface and/or other decals. All images in
+    /// the scene must use the same sampler.
```
```suggestion
    /// blended with the underlying surface and/or other decals. All images in
    /// the scene must use the same sampler.
```

Please reword this so that it's clear that all images used for _decals_ in the scene must use the same sampler (descriptor?), but other images can use other samplers.

pcwalton replied:
Good catch, done.

JMS55 commented on crates/bevy_pbr/src/decal/clustered.wgsl:
```diff
@@ -0,0 +1,178 @@
+) -> vec4<f32> {
+    var base_color = initial_base_color;
+
+#ifdef CLUSTERED_DECALS_ARE_USABLE
```
```suggestion
#ifdef CLUSTERED_DECALS_ARE_USABLE
```

Is this supposed to be called that? Do you mean to call it "CLUSTERED_DECALS_USED_IN_SCENE" or something?

pcwalton replied:
Yeah, it's supposed to be called `CLUSTERED_DECALS_ARE_USABLE`; it's `#define`'d based on whether bindless is available.

djeedai replied:
`HAS_CLUSTERED_DECALS` or `WITH_CLUSTERED_DECALS` are more common

JMS55 reviewed (approved):
Generally looks fine. Two requests:
1. Can you unify the naming a little more? In some places they're called clustered decals, othertimes projected. I'd rather stick with clustered everywhere.
2. Won't block on it, but I would like to have one of the decals in the example be a custom material, so that users can see how to do that. I'm not sure myself, I think you'd need a switch statement on the decal type for any material on a surface that could receive a decal? Not sure.

pcwalton commented:
@JMS55 Do you still think a projector should be called a `DecalProjector`? Or should I change that to `ClusteredDecal`? I took the name from [Unity's Decal Projectors](https://docs.unity3d.com/Packages/com.unity.render-pipelines.universal@12.0/manual/renderer-feature-decal.html#decal-projector-component).

alice-i-cecile commented:
Merging #16600 first since it's ready; we can merge the examples together into one here :)

pcwalton commented:
@JMS55 I updated this with a couple of important changes:
1. The term "projector" has been replaced with "clustered" everywhere.
2. There's now a `tag`, which is intended to be used for application-specific purposes like indexing an array of normal maps. I added a simple demonstration of its use to the example.

pcwalton commented on crates/bevy_pbr/src/cluster/assign.rs:
```diff
@@ -249,6 +257,19 @@ pub(crate) fn assign_objects_to_clusters(
     }
 
+    // TODO: this should be "if supports_bindless" or something
+    if binding_arrays_are_usable(&render_device, &render_adapter) {
```
I removed the out-of-date comment. It should already be pulled out of the loop.

rparrett reviewed (approved):
Limited review, most of the rendering code is out of my league.

tychedelia commented on crates/bevy_pbr/src/decal/clustered.wgsl:
```diff
@@ -0,0 +1,183 @@
+//
+// Returns true if another decal was found or false if no more decals were found
+// for this position.
+fn clustered_decal_iterator_next(iterator: ptr<function, ClusteredDecalIterator>) -> bool {
```
cool pattern, hadn't seen this before

tychedelia reviewed (approved):
Tested on m1 mbp.

pcwalton commented on crates/bevy_pbr/src/decal/clustered.rs:
```diff
@@ -0,0 +1,369 @@
+//! surface within its bounds along the positive Z axis. In Bevy, clustered
+//! decals use the *clustered forward* rendering technique.
+//!
+//! Clustered decals are the highest-quality types of decals that Bevy supports,
```
Yeah, we support forward decals as well.

pcwalton commented on crates/bevy_pbr/src/decal/clustered.rs:
```diff
@@ -0,0 +1,369 @@
+/// The maximum number of decals that can be present in a view.
+///
+/// This number is currently relatively low in order to work around the lack of
+/// first-class binding arrays in `wgpu`. When that feature is implemented, this
```
There's https://github.com/gfx-rs/wgpu/issues/3637, is that what you mean?

pcwalton commented on crates/bevy_pbr/src/decal/clustered.rs:
```diff
@@ -0,0 +1,369 @@
+//!
+//! Clustered decals are the highest-quality types of decals that Bevy supports,
+//! but they require bindless textures. This means that they presently can't be
+//! used on WebGL 2, WebGPU, macOS, or iOS. Bevy's clustered decals can be used
```
For WebGL, that's expected. I don't think there's much point in gating off decals with an ifdef in the example's custom shader if the example requires a platform that decals don't work on to begin with.
For WebGPU, the indirect parameters metadata error is unrelated and I'd rather not fix that here.

pcwalton replied:
Good point on the macOS/iOS thing. They don't have *true* bindless, but I forget that wgpu has a partial workaround that's good enough to make clustered decals work. I removed the mention of those platforms.

djeedai replied:
The error means some bound buffer is marked read-only but contains some atomic. That's not allowed by WGSL. In Hanabi I use some #ifdef to change the definition of the atomic fields to non-aromic and read them as non-atomic where it's safe to do so, to avoid that error. That's a bit tedious but it's clean.

pcwalton commented on examples/3d/clustered_decals.rs:
```diff
@@ -0,0 +1,541 @@
+            Vec3::ZERO,
+            Vec2::splat(1.1),
+        ))
+        .insert(Selection::DecalA);
```
Ok, changed all of them to tuples.

alice-i-cecile commented:
Example runner looks good! https://pixel-eagle.com/project/B25A040A-A980-4602-B90C-D480AB84076D/run/7475/compare/7463

alice-i-cecile commented:
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1986 if you'd like to help out.
